### PR TITLE
ARTESCA-4300 raise alert when pods get OOMKilled

### PR DESCRIPTION
**Pull request by aprucolimartins**
**Component**:

**Context**: 

**Summary**:

**Acceptance criteria**: 

---

Closes: #ISSUE_NUMBER

### CODE REVIEW

File: salt/metalk8s/addons/prometheus-operator/deployed/prometheus-rules.sls [hunk 1]
```diff
@@ -281,4 +281,52 @@ spec:
...
+          * on (namespace, pod) group_left(node) kube_pod_info
+        ) > 0
+      for: 1s
+      labels:
+        severity: warning
+    - alert: KubeContainerOOMKillSurge
+      annotations:
+        description: 'OOMKill Surge: Container {{ $labels.namespace }}/{{ $labels.pod }}/{{ $labels.container }}
+          in Node {{ $labels.node }} caused Pod restart by have been exceed its memory allocation 
+          limit (OOMKilled). It accumulates {{ $value }} restarts in last 5 minutes.'
```
Comment (TeddyAndrieux) [inline salt/metalk8s/addons/prometheus-operator/deployed/prometheus-rules.sls]: Here it's not "$value restarts" but more "$value restarts because of OOM kill, I think it's important to mention.

Btw I wonder if we want this other rule, just counting the number of OOM kill restarts is enough to me, because a pod may restart for good reason, so catching the total number of restarts for an alert is not necessarily interesting (even if we filter on OOMKilled at the beginning, it means in the alert message you will have the total number of the restart when maybe all of them are "legit" except the OOMKilled one)

I would suggest going only with a warning alert when there is an OOMKilled and to do so I think we do not need the `kube_pod_container_status_last_terminated_reason` metrics, we can handle it directly with `container_oom_events_total`

Comment (aprucolimartins) [reply salt/metalk8s/addons/prometheus-operator/deployed/prometheus-rules.sls]: As explained in this comment ( https://scality.atlassian.net/browse/ARTESCA-4300?focusedCommentId=346858 ) the metric `container_oom_events_total` is capable to indicates OOMKill events for a period of 5 minutes.
My proposition ( https://scality.atlassian.net/browse/ARTESCA-4300?focusedCommentId=346867 ) is to raise 2 distinct alerts : a CRITICAL one, based on `container_oom_events_total` to indicate OOMKill surge and a WARNING that will stay active still concerned pods restarts.

Comment (aprucolimartins) [reply salt/metalk8s/addons/prometheus-operator/deployed/prometheus-rules.sls]: To close discussion, the container_oom_events_total metric is bugged: https://github.com/google/cadvisor/issues/3015

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+metadata:
+  labels:
+    app: prometheus-operator
+    app.kubernetes.io/managed-by: salt
+    app.kubernetes.io/name: prometheus-operator
+    app.kubernetes.io/part-of: metalk8s
+    chart: prometheus-operator-8.1.2
+    metalk8s.scality.com/monitor: ''
+    heritage: metalk8s
+  name: prometheus-operator-container-exporter
```
Comment (TeddyAndrieux) [inline salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: I know it was the case for the prometheus rules already defined, but to me those labels are wrong, this manifest is not part of the prometheus-operator chart (and the version is not the one we use today so ...)

It's not linked to your PR, but for this new one could you change a bit the labels and name
Something like that

```suggestion
    app.kubernetes.io/part-of: metalk8s
    metalk8s.scality.com/monitor: ''
  name: metalk8s-containers.rules
```

(feel free to update the "node-exporter" rules as well :) )

File: buildchain/buildchain/salt_tree.py [hunk 3]
```diff
@@ -383,7 +383,14 @@ def task(self) -> types.TaskDict:
...
     Path("salt/metalk8s/addons/prometheus-operator/deployed/namespace.sls"),
-    Path("salt/metalk8s/addons/prometheus-operator/deployed/prometheus-rules.sls"),
+    Path(
+        "salt/metalk8s/addons/prometheus-operator/deployed/",
+        "node-exporter-alerts-rules.sls",
+    ),
+    Path(
+        "salt/metalk8s/addons/prometheus-operator/deployed/",
+        "container-exporter-alerts-rules.sls",
+    ),
```
Comment (TeddyAndrieux) [inline buildchain/buildchain/salt_tree.py]: Not entirely sure about the names, I would suggest just `container-alert-rules.sls` and `node-alert-rules.sls`

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        summary: 'Pod killed by exceeding memory allocation limits (OOMKilled).'
+      expr: |-
+        max_over_time(
+          (
+            kube_pod_container_status_last_terminated_reason{job="kube-state-metrics", namespace=~".*", reason="OOMKilled"}
+            * on (namespace, pod, container) kube_pod_container_status_restarts_total
+            * on (namespace, pod) group_left(node) kube_pod_info
+          ) [3d:]
+        ) > 0
+      for: 1s
```
Comment (TeddyAndrieux) [inline salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: The `for` could be removed to me here

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        round(
+          delta(
+            (
+              kube_pod_container_status_last_terminated_reason{job="kube-state-metrics", namespace=~".*", reason="OOMKilled"}
+              * on (namespace, pod, container) kube_pod_container_status_restarts_total
+              * on (namespace, pod) group_left(node) kube_pod_info
+            ) [61m:]
+          ), 1
+        ) > 0
+      for: 1s
```
Comment (TeddyAndrieux) [inline salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: ditto

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  namespace: metalk8s-monitoring
+spec:
+  groups:
+  - name: container-exporter
+    rules:
+    - alert: KubeContainerOOMKilled
+      annotations:
+        description: 'Container {{ $labels.namespace }}/{{ $labels.pod }}/{{ $labels.container }}
+          in Node {{ $labels.node }} caused Pod restart by have been exceed its memory allocation
+          limit (OOMKilled). This container accumulates a total of {{ printf "%.0f" $value }} restart(s).'
```
Comment (TeddyAndrieux) [inline salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: I would re-phrase it a bit

```suggestion
        description: 'Container {{ $labels.namespace }}/{{ $labels.pod }}/{{ $labels.container }}
          on Node {{ $labels.node }} exceeds its memory allocation limit (OOMKilled).
          This container accumulates a total of {{ printf "%.0f" $value }} restart(s).'
```

What do you think ?

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        summary: 'Pod killed by exceeding memory allocation limits (OOMKilled).'
+      expr: |-
+        max_over_time(
+          (
+            kube_pod_container_status_last_terminated_reason{job="kube-state-metrics", namespace=~".*", reason="OOMKilled"}
+            * on (namespace, pod, container) kube_pod_container_status_restarts_total
+            * on (namespace, pod) group_left(node) kube_pod_info
+          ) [3d:]
```
Comment (TeddyAndrieux) [inline salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: I'm not sure to get why you put `[3d:]` instead of `[3d]` ?

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+spec:
+  groups:
+  - name: container-exporter
+    rules:
+    - alert: KubeContainerOOMKilled
+      annotations:
+        description: 'Container {{ $labels.namespace }}/{{ $labels.pod }}/{{ $labels.container }}
+          in Node {{ $labels.node }} caused Pod restart by have been exceed its memory allocation
+          limit (OOMKilled). This container accumulates a total of {{ printf "%.0f" $value }} restart(s).'
+        summary: 'Pod killed by exceeding memory allocation limits (OOMKilled).'
```
Comment (TeddyAndrieux) [inline salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: To me it's not the Pod that get killed but just the container

**Review by TeddyAndrieux — COMMENTED** [review]
Almost good to me, mostly nitpicking and rewording 

+ You will need to re-extract rules using the rule_extractor (feel free to reach me directly if you need help about it)

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls (hunk 7 above)
Comment (aprucolimartins) [reply salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: [3d] don't work in subqueries. So it's [3d:].
https://prometheus.io/blog/2019/01/28/subquery-support/#examples

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls (hunk 8 above)
Comment (aprucolimartins) [reply salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: I agree you. But kill a container causes pod restart.
We need just find a better way to describe it.

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls (hunk 4 above)
Comment (aprucolimartins) [reply salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: Done !!!

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls (hunk 5 above)
Comment (aprucolimartins) [reply salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: Done !!!

File: buildchain/buildchain/salt_tree.py (hunk 3 above)
Comment (aprucolimartins) [reply buildchain/buildchain/salt_tree.py]: Done !!!

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls (hunk 8 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: Not really, if you have a Pod with 2 containers, and one of them get killed to me it's the only one to restart, the other one will not restart

File: salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls (hunk 7 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/addons/prometheus-operator/deployed/container-exporter-alerts-rules.sls]: Oh ok good to know :+1:

File: salt/metalk8s/addons/prometheus-operator/deployed/node-alerts-rules.sls [hunk 9]
```diff
@@ -17,13 +17,8 @@ apiVersion: monitoring.coreos.com/v1
...
 metadata:
   labels:
-    app: prometheus-operator
-    app.kubernetes.io/managed-by: salt
-    app.kubernetes.io/name: prometheus-operator
     app.kubernetes.io/part-of: metalk8s
-    chart: prometheus-operator-8.1.2
     metalk8s.scality.com/monitor: ''
-    heritage: metalk8s
   name: prometheus-operator-node-exporter
```
Comment (TeddyAndrieux) [inline salt/metalk8s/addons/prometheus-operator/deployed/node-alerts-rules.sls]: (I would rename it also :relaxed: )

File: salt/metalk8s/addons/prometheus-operator/deployed/container-alert-rules.sls [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+          ) [3d:]
+        ) > 0
+      labels:
+        severity: warning
+    - alert: KubeContainerOOMKillSurge
+      annotations:
+        description: 'OOMKill Surge: Container {{ $labels.namespace }}/{{ $labels.pod }}/{{ $labels.container }}
+          in Node {{ $labels.node }} caused Pod restart by have been exceed its memory allocation
+          limit (OOMKilled). This container restarted {{ printf "%.0f" $value }} time(s) in last hour.'
+        summary: 'OOMKill Surge: Pod killed by exceeding memory allocation limits (OOMKilled) in last hour.'
```
Comment (TeddyAndrieux) [inline salt/metalk8s/addons/prometheus-operator/deployed/container-alert-rules.sls]: I would rephrase it the same as the other one

File: salt/metalk8s/addons/prometheus-operator/deployed/init.sls [hunk 11]
```diff
@@ -7,5 +7,6 @@ include:
   - .dashboards
   - .service-configuration
   - .chart
-  - .prometheus-rules
+  - .node-exporter-alerts-rules
+  - .container-exporter-alerts-rules
```
Comment (TeddyAndrieux) [inline salt/metalk8s/addons/prometheus-operator/deployed/init.sls]: You will need to rename it here as well

File: salt/metalk8s/addons/prometheus-operator/deployed/container-alert-rules.sls (hunk 10 above)
Comment (aprucolimartins) [reply salt/metalk8s/addons/prometheus-operator/deployed/container-alert-rules.sls]: Done !!!

File: salt/metalk8s/addons/prometheus-operator/deployed/init.sls (hunk 11 above)
Comment (aprucolimartins) [reply salt/metalk8s/addons/prometheus-operator/deployed/init.sls]: Done !!!

File: salt/metalk8s/addons/prometheus-operator/deployed/node-alerts-rules.sls (hunk 9 above)
Comment (aprucolimartins) [reply salt/metalk8s/addons/prometheus-operator/deployed/node-alerts-rules.sls]: Done !!!

**Review by TeddyAndrieux — DISMISSED** [review]

**Review by eg-ayoub — APPROVED** [review]
